src/Core/Batch/BatchTrait.php

```diff
@@ -0,0 +1,195 @@
+                    memory_get_usage()
+                )
+            );
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns an array representation of a callback which will be used to write
+     * batch items.
+     *
+     * @return array
+     */
+    protected abstract function getCallback();
+
+    /**
+     * @param array $options [optional] {
+     *     Configuration options.
+     *
+     *     @type resource $debugOutputResource A resource to output debug output
+     *           to.
+     *     @type bool $debugOutput Whether or not to output debug information.
+     *           **Defaults to** false.
+     *     @type array $batchOptions A set of options for a BatchJob.
+     *           {@see \Google\Cloud\Core\Batch\BatchJob::__construct()} for
+     *           more details.
+     *           **Defaults to** ['batchSize' => 1000,
+     *                            'callPeriod' => 2.0,
+     *                            'workerNum' => 2].
+     *     @type array $clientConfig A config used to construct the client upon
+     *           which requests will be made.
+     *     @type BatchRunner $batchRunner A BatchRunner object. Mainly used for
+     *           the tests to inject a mock. **Defaults to** a newly created
+     *           BatchRunner.
+     *     @type string $identifier An identifier for the batch job.
+     *     @type string $batchMethod The name of the batch method used to
+     *           deliver items.
+     * }
+     * @throws \InvalidArgumentException
+     */
+    private function setCommonBatchProperties(array $options)
+    {
+        if (!isset($options['identifier'])) {
+            throw new \InvalidArgumentException(
+                'A valid identifier is required in order to register a job.'
+            );
+        }
+
+        if (!isset($options['batchMethod'])) {
+            throw new \InvalidArgumentException(
+                'A batchMethod is required.'
+            );
+        }
+
+        $this->batchMethod = $options['batchMethod'];
+        $this->identifier = $options['identifier'];
+        $this->debugOutputResource = isset($options['debugOutputResource'])
+            ? $options['debugOutputResource']
+            : null;
+        $this->debugOutput = isset($options['debugOutput'])
+            ? $options['debugOutput']
+            : false;
+        $this->clientConfig = isset($options['clientConfig'])
+            ? $options['clientConfig']
+            : [];
+        $batchOptions = isset($options['batchOptions'])
+            ? $options['batchOptions']
+            : [];
+        $this->batchOptions = $batchOptions + [
+            'batchSize' => 1000,
+            'callPeriod' => 2.0,
+            'workerNum' => 2
+        ];
+        $this->batchRunner = isset($options['batchRunner'])
+            ? $options['batchRunner']
+            : new BatchRunner();
+        $this->batchRunner->registerJob(
+            $this->identifier,
+            [$this, 'send'],
```

Comment:
I just finished up a little testing of my own using the examples provided above. Both sets of code returned very similar performance results for me (I tested the latest code removing `send` and `getCallback` as well, and saw no difference). I'll try to run some more in a little bit to see if these results were an anomaly.

> The minimum bound of the performance we're getting with the current code is acceptable, so maybe we can move forward?

I would hate to introduce something in this PR that diminishes performance - but I do think these benchmarks are an extreme use case and most users won't be affected too harshly. I would be okay with moving forward if you are as well!

Using an older version ([this commit](https://github.com/GoogleCloudPlatform/google-cloud-php/commit/81319933b4c914d5b4bb00c88712c07437c73822)):
```
$ ab -n 10000 -c 20 localhost:8123/ex.php
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            8123

Document Path:          /ex.php
Document Length:        51 bytes

Concurrency Level:      20
Time taken for tests:   176.947 seconds
Complete requests:      10000
Failed requests:        1233
   (Connect: 0, Receive: 0, Length: 1233, Exceptions: 0)
Total transferred:      1618765 bytes
HTML transferred:       508765 bytes
Requests per second:    56.51 [#/sec] (mean)
Time per request:       353.893 [ms] (mean)
Time per request:       17.695 [ms] (mean, across all concurrent requests)
Transfer rate:          8.93 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       4
Processing:    20  353  39.3    354    1128
Waiting:       19  352  39.3    353    1127
Total:         20  353  39.3    354    1128

Percentage of the requests served within a certain time (ms)
  50%    354
  66%    359
  75%    361
  80%    363
  90%    368
  95%    373
  98%    378
  99%    385
 100%   1128 (longest request)
```

And the code introduced in this PR ([commit](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/534/commits/5a9c89d5a1005092017860ab89d6dd0468ee2b19)):
```
$ ab -n 10000 -c 20 localhost:8123/ex.php
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            8123

Document Path:          /ex.php
Document Length:        51 bytes

Concurrency Level:      20
Time taken for tests:   176.260 seconds
Complete requests:      10000
Failed requests:        7717
   (Connect: 0, Receive: 0, Length: 7717, Exceptions: 0)
Total transferred:      1627711 bytes
HTML transferred:       517711 bytes
Requests per second:    56.73 [#/sec] (mean)
Time per request:       352.520 [ms] (mean)
Time per request:       17.626 [ms] (mean, across all concurrent requests)
Transfer rate:          9.02 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       4
Processing:    20  352  21.8    352     669
Waiting:       19  351  21.8    351     668
Total:         20  352  21.8    352     669

Percentage of the requests served within a certain time (ms)
  50%    352
  66%    358
  75%    361
  80%    363
  90%    368
  95%    372
  98%    378
  99%    382
 100%    669 (longest request)
```